### PR TITLE
Adds +INIT and $OS_RELEASE to cache id

### PR DESCRIPTION
**Pull request by idelvall**
This PR:
- Includes the distro information (from `/etc/os-release`) in the cache id default value. This avoids the scenario of reusing the cache entries across potentially incompatibles shared libraries.
- Includes a `+INIT` UDC used for initialize the build environment for the rest UDCs to be invoked. 
- Exports the cache id to the build environment as `$CARGO_HOME_CACHE_ID`

### CODE REVIEW

File: rust/README.md [hunk 1]
```diff
@@ -44,7 +56,10 @@ Use this argument when you want to `SAVE ARTIFACT` from the target folder (mount
 
 For example `--output="release/[^\./]+"` would keep all the files in `/target/release` that don't have any extension.
 
-### Examples:
+### Thread safety
+This UDC should is thread safe. Parallel builds of targets calling this UDC should be free of race conditions.
```
Comment (vladaionescu) [inline rust/README.md:60]:
```suggestion
This UDC is thread safe. Parallel builds of targets calling this UDC should be free of race conditions.
```

File: rust/Earthfile [hunk 2]
```diff
@@ -1,24 +1,45 @@
...
 #   - output: Regex to match the files within the target folder to be copied from the cache to the caller filesystem (image layers).
 #     Use this argument when you want to SAVE an ARTIFACT from the target folder (mounted cache), always trying to minimize the total size of the copied fileset.
 #     For example --output="release/[^\./]+" would keep all the files in /target/release that don't have any extension.
 CARGO:
     COMMAND
     ARG --required args
-    ARG keep_fingerprints=false
-    ARG sweep_days=4
+    ARG keep_fingerprints=$(cat /earthly/cfg/keep_fingerprints)
+    ARG sweep_days=$(cat /earthly/cfg/sweep_days)
```
Comment (vladaionescu) [inline rust/Earthfile:38]: Possible idea - I wonder if we should detect if `rust+INIT` was called and call it implicitly here if not.

**Review by vladaionescu — APPROVED** [review]

File: rust/Earthfile [hunk 3]
```diff
@@ -1,24 +1,45 @@
...
+#   - sweep_days (4): +CARGO calls use cargo-sweep to clean build artifacts that haven't been accessed for this number of days.
+INIT:
+    COMMAND
+    DO +INSTALL_CARGO_SWEEP
+    RUN mkdir -p /earthly/cfg
+
+    # cache_id
+    ARG EARTHLY_TARGET_PROJECT_NO_TAG
+    ARG OS_RELEASE=$(md5sum /etc/os-release | cut -d ' ' -f 1)
+    ARG cache_id="${EARTHLY_TARGET_PROJECT_NO_TAG}#${OS_RELEASE}#earthly-cargo-cache"
```
Comment (idodod) [inline rust/Earthfile:15]: You might want to use `SET` here instead of `ARG`, which will indicate this is not something that can/should be overridden (unless I'm misunderstanding it).

File: rust/Earthfile [hunk 4]
```diff
@@ -60,8 +81,8 @@ get-tomljson:
     FROM alpine:3.18.3
     ARG USERARCH
     ARG version=2.1.0
-    RUN wget -O tomljson.tar.xz https://github.com/pelletier/go-toml/releases/download/v${version}/tomljson_${version}_linux_${USERARCH}.tar.xz && \
-        tar -xf tomljson.tar.xz && \
+    RUN wget -O tomljson.tar.xz https://github.com/pelletier/go-toml/releases/download/v${version}/tomljson_${version}_linux_${USERARCH}.tar.xz; \
+        tar -xf tomljson.tar.xz; \
```
Comment (idodod) [inline rust/Earthfile:85]: is this change intentional or possibly this PR didn't included latest main (with Alex's changes)?

File: rust/Earthfile [hunk 5]
```diff
@@ -70,43 +91,41 @@ get-jq:
...
 INSTALL_CARGO_SWEEP:
     COMMAND
-    DO +RUN_WITH_CACHE --command="set -e;
-        if [ ! -f \$CARGO_HOME/bin/cargo-sweep ]; then
-          cargo install cargo-sweep --root \$CARGO_HOME;
-        fi;"
+    RUN if [ ! -f $CARGO_HOME/bin/cargo-sweep ]; then \
+          echo "Installing cargo sweep" ; \
+          cargo install cargo-sweep --root $CARGO_HOME; \
+        fi;
```
Comment (idodod) [inline rust/Earthfile:104]: just curious why you used a `RUN if` instead of `IF ... RUN`

File: rust/Earthfile (hunk 4 above)
Comment (idelvall) [reply rust/Earthfile:85]: it wasn't, thanks!

File: rust/Earthfile (hunk 5 above)
Comment (idelvall) [reply rust/Earthfile:104]: `IF ... RUN` runs two shell instances also, within RUN we just have bash semantics (for the conditional), that I am more familiarized with

File: rust/Earthfile [hunk 6]
```diff
@@ -1,24 +1,45 @@
...
+    COMMAND
+    DO +INSTALL_CARGO_SWEEP
+    RUN mkdir -p /earthly/cfg
+
+    # cache_id
+    ARG EARTHLY_TARGET_PROJECT_NO_TAG
+    ARG OS_RELEASE=$(md5sum /etc/os-release | cut -d ' ' -f 1)
+    ARG cache_id="${EARTHLY_TARGET_PROJECT_NO_TAG}#${OS_RELEASE}#earthly-cargo-cache"
+    RUN echo "$cache_id">/earthly/cfg/cache_id
+    ENV CARGO_HOME_CACHE_ID=$cache_id
```
Comment (xv-ian-c) [inline rust/Earthfile:17]: This is nice as a workaround but can I suggest you consider making this a first class feature of Earthly UDCs. e.g.

```
IMPORT github.com/earthly/lib/rust AS rust-udc WITH --foo=bar --cache-id=foo
```
To set some `ARG --global` declared at the top of the UDC?

File: rust/Earthfile (hunk 4 above)
Comment (xv-ian-c) [reply rust/Earthfile:85]: Is this about `&&` vs `;` or something else? 

In the absence of `set -e` I think `&&` ought to be preferred.

File: rust/Earthfile [hunk 7]
```diff
@@ -70,43 +91,41 @@ get-jq:
...
     SAVE ARTIFACT jq
 
 # INSTALL_CARGO_SWEEP installs cargo-sweep if it doesn't exist already.
 INSTALL_CARGO_SWEEP:
     COMMAND
-    DO +RUN_WITH_CACHE --command="set -e;
-        if [ ! -f \$CARGO_HOME/bin/cargo-sweep ]; then
-          cargo install cargo-sweep --root \$CARGO_HOME;
-        fi;"
+    RUN if [ ! -f $CARGO_HOME/bin/cargo-sweep ]; then \
```
Comment (xv-ian-c) [inline rust/Earthfile:101]: Why no longer caching this?

File: rust/Earthfile (hunk 2 above)
Comment (xv-ian-c) [reply rust/Earthfile:38]: And if not I think every `COMMAND` which depends on this having been run ought to say so in it's doc comment?

Comment (xv-ian-c) [reply rust/Earthfile:38]: And the doc file itself says "It is meant to be called once per build environment..." which I think maybe ought to be "It must be called..."?

File: rust/README.md [hunk 8]
```diff
@@ -2,40 +2,52 @@
...
 
-Cargo caches compilations of packages in `target` folder based on their last modification timestamps. 
+### Arguments
+#### `cache_id`
+Overrides default ID of the global `$CARGO_HOME` cache. Its value is exported to the build environment under the entry: `$CARGO_HOME_CACHE_ID`.
 
+#### `keep_fingerprints (false)`
+Instructs the following `+CARGO` calls to don't remove the Cargo fingerprints of the source packages. Use only when source packages have been COPYed with `--keep-ts `option.
+Cargo caches compilations of packages in `target` folder based on their last modification timestamps.
 By default, this UDC removes the fingerprints of the packages found in the source code, to force their recompilation and work even when the Earthly `COPY` commands used overwrote the timestamps.
```
Comment (xv-ian-c) [inline rust/README.md:31]: TBH I'm not sure how well this is working, I think there should be a strong recommendation to use `--keep-ts` and to enable `keep_fingerprints` (clarifying the double negative: I mean that the fingerprints should be left alone).

File: rust/Earthfile (hunk 2 above)
Comment (idelvall) [reply rust/Earthfile:38]: I added checks to make it clear it has to be called once. Calling implicitly could make installing dependencies multiple times (cargo sweep)

File: rust/Earthfile (hunk 3 above)
Comment (idelvall) [reply rust/Earthfile:15]: Good point, I tried. But found that `--arg-scope-and-set` needs to be set in the calling Earthfile (that seems like a bug), and also builtin args cannot be set via `LET` (that looks more appropriate), so decide to leave it for now

File: rust/Earthfile (hunk 6 above)
Comment (idelvall) [reply rust/Earthfile:17]: This is a good feature to add for sure Ian, mind creating a ticket for it? 
https://github.com/earthly/earthly/issues/new?assignees=&labels=type%3Aproposal&projects=&template=feature-proposal.md&title=

File: rust/README.md (hunk 8 above)
Comment (idelvall) [reply rust/README.md:31]: Agree. We will be improving the way timestamps are handled soon, and this hack will be removed.

File: rust/Earthfile (hunk 7 above)
Comment (idelvall) [reply rust/Earthfile:101]: Before it was also being installed in the build environment (notice the `$CARGO_INSTALL_ROOT` override).

This change moves it to +INIT so, the user can run this before copying files from their context, and this installation is cached even when source files change.

The reason to move this out of the cache was to avoid the scenario of BK garbage collecting these files, while the RUN command installing it still kept cached. 

We could have handled this case by always checking the command exists before using it (in the same RUN command) but overall I thought that moving the installation to the init phase would make things cleaner.

File: rust/Earthfile (hunk 3 above)
Comment (idelvall) [reply rust/Earthfile:19]: @vladaionescu any thoughts on this?

File: rust/Earthfile (hunk 6 above)
Comment (xv-ian-c) [reply rust/Earthfile:21]: https://github.com/earthly/earthly/issues/3488

File: rust/Earthfile (hunk 3 above)
Comment (idelvall) [reply rust/Earthfile:19]: https://github.com/earthly/earthly/issues/3490

Comment (vladaionescu) [reply rust/Earthfile:19]: Leaving it SGTM
